### ISSUE TITLE
Failed to build package gettext for target x86_64-unknown-cygwin

### ISSUE THREAD

**Issue by Kaos-Industries** [mxe/mxe#2316]
Sorry that this is my third issue for the same compile session, but I just had the command that had already taken 4 hours fail, and seemingly over a trivial package for a target I don't even need. 

Any idea why?

Here's the [relevant log](https://gist.github.com/Kaos-Industries/2d8584c8ab70dfdc94847b75f3dcd2bc), or at least what I think is.

Would really appreciate some help on this.

**Comment by tonytheodore** [comment]
Can you check if https://github.com/mxe/mxe/pull/2317 fixes it?

**Comment by Kaos-Industries** [comment]
Thanks for your prompt replies on this, it's really appreciated. 

Unfortunately I'm pretty experienced when it comes to both Git and patches - how exactly would I apply #2317?

**Comment by tonytheodore** [comment]
In every PR, there's always a link to the command line instructions:
```
git checkout -b tonytheodore-gettext-cygwin
git pull https://github.com/tonytheodore/mxe.git gettext-cygwin
```

**Comment by Kaos-Industries** [comment]
I couldn't find that command anywhere in the PR and still can't unfortunately. Whereabouts is it usually shown?

Anyway, thanks for the commands, I'll apply the patch now and test again.

**Comment by Kaos-Industries** [comment]
Unfortunately, it still seems to fail:

```
Failed to build package gettext for target x86_64-unknown-cygwin!
------------------------------------------------------------
make[3]: *** [Makefile:414: all-recursive] Error 1
make[3]: Leaving directory '/usr/mxe/tmp-gettext-x86_64-unknown-cygwin/gettext-0.19.8.1.build_'
make[2]: *** [Makefile:370: all] Error 2
make[2]: Leaving directory '/usr/mxe/tmp-gettext-x86_64-unknown-cygwin/gettext-0.19.8.1.build_'
make[1]: *** [Makefile:832: build-only-gettext_x86_64-unknown-cygwin] Error 2
make[1]: Leaving directory '/usr/mxe'
real    58m21.585s
user    10m41.470s
sys     24m51.379s
------------------------------------------------------------
[log]      /usr/mxe/log/gettext_x86_64-unknown-cygwin

make: *** [Makefile:820: /usr/mxe/usr/x86_64-unknown-cygwin/installed/gettext] Error 1
make: *** Waiting for unfinished jobs....
make: *** [Makefile:820: /usr/mxe/usr/x86_64-w64-mingw32.static/installed/gcc] Interrupt
```

This was the command that I ran:

`make MXE_TARGETS="x86_64-w64-mingw32.static" lame libass libvpx ogg opus sdl2 theora vorbis x264 -j$(nproc) MXE_USE_CCACHE=`

I tried both doing `checkout` and `pull` in the directory of the first failed install, and deleting and re-cloning MXE and then doing `checkout` and `pull` - both failed.

**Comment by tonytheodore** [comment]
> I couldn't find that command anywhere in the PR and still can't unfortunately. Whereabouts is it usually shown?

<img width="780" alt="Screen Shot 2019-03-31 at 10 58 50" src="https://user-images.githubusercontent.com/537773/55283000-8b667380-53a4-11e9-9548-39b7a64e359c.png">

**Comment by tonytheodore** [comment]
> This time I can't even find a relevant log

The log location is shown at the bottom of the error message:
```
[log]      /usr/mxe/log/gettext_x86_64-unknown-cygwin
```

Can you try building gettext in a normal cygwin session? If you can get that working, then we can fix mxe accordingly.

**Comment by Kaos-Industries** [comment]
Yeah, as I suspected, the command line instructions only seem to display if you have write access to the repo. This is what I see:

![Screenshot_2019-03-31 gettext fix build on cygwin by tonytheodore · Pull Request #2317 · mxe mxe](https://user-images.githubusercontent.com/16962031/55283251-65ae7f80-534e-11e9-8597-358bed30c811.png)

>The log location is shown at the bottom of the error message:
>[log]      /usr/mxe/log/gettext_x86_64-unknown-cygwin

That file seems to be a binary .SYS file rather than anything I can read in a text editor. I've uploaded it to MediaFire if you still need it:

http://www.mediafire.com/file/zgkuvny9453zpif/gettext_x86_64-unknown-cygwin/file

> Can you try building gettext in a normal cygwin session? If you can get that working, then we can fix mxe accordingly.

I assume you mean from source rather than just installing the package via Cygwin's package? If the former, then I'm currently working on it and will get back to you.

**Comment by Kaos-Industries** [comment]
Build of the latest version of gettext under Cygwin just failed with the following error:

`libtool: link: ranlib .libs/libgettextlib.a
libtool: link: rm -fr .libs/libgettextlib.lax
/usr/bin/sed: can't read (3)/gettext-0.19.8/gettext-tools/intl/libintl.la: No such file or directory
libtool:   error: '(3)/gettext-0.19.8/gettext-tools/intl/libintl.la' is not a valid libtool archive
make[5]: *** [Makefile:2868: libgettextlib.la] Error 1
make[5]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/New folder (3)/gettext-0.19.8/gettext-tools/gnulib-lib'
make[4]: *** [Makefile:2618: all] Error 2
make[4]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/New folder (3)/gettext-0.19.8/gettext-tools/gnulib-lib'
make[3]: *** [Makefile:2026: all-recursive] Error 1
make[3]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/New folder (3)/gettext-0.19.8/gettext-tools'
make[2]: *** [Makefile:1892: all] Error 2
make[2]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/New folder (3)/gettext-0.19.8/gettext-tools'
make[1]: *** [Makefile:413: all-recursive] Error 1
make[1]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/New folder (3)/gettext-0.19.8'
make: *** [Makefile:369: all] Error 2`

I tried uninstalling and reinstalling `libtool` and running the build again but the error still persists.

**Comment by tonytheodore** [comment]
Ah, it makes sense re the PR and command line instructions.

The log file is a symlink to a date-stamped sub-directory in mxe/log, you should be able to find the log in the latest folder there. I don't use cygwin but these files shouldn't be hidden in any way.

> Build of the latest version of gettext under Cygwin just failed with the following error:
```
/usr/bin/sed: can't read (3)/gettext-0.19.8/gettext-tools/intl/libintl.la: No such file or directory
```

building in a directory with spaces (e.g. `New folder (3)`) is unlikely to work reliably.

**Comment by Kaos-Industries** [comment]
I'd looked in the dated folder but didn't notice any log files that weren't download-related, but just checked again and it is indeed there. [Here](https://gist.github.com/Kaos-Industries/b7ccde3f4c2052aadbe96819cdb65ab1) it is.

Interesting, I didn't think UNIX-like systems had such limitations. Renamed the folder and trying to build again now.

**Comment by Kaos-Industries** [comment]
Second compile of gettext on Cygwin also failed with an error, although it seems to be a different one:

```
/libgettextsrc.dll.a
.libs/libgettextsrc_la-locating-rule.o:locating-rule.c:(.rdata$.refptr.xmlFree[.refptr.xmlFree]+0x0): undefined reference to `xmlFree'
collect2: error: ld returned 1 exit status
make[5]: *** [Makefile:2635: libgettextsrc.la] Error 1
make[5]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/gettext/gettext-0.19.8/gettext-tools/src'
make[4]: *** [Makefile:2556: all] Error 2
make[4]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/gettext/gettext-0.19.8/gettext-tools/src'
make[3]: *** [Makefile:2026: all-recursive] Error 1
make[3]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/gettext/gettext-0.19.8/gettext-tools'
make[2]: *** [Makefile:1892: all] Error 2
make[2]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/gettext/gettext-0.19.8/gettext-tools'
make[1]: *** [Makefile:413: all-recursive] Error 1
make[1]: Leaving directory '/cygdrive/b/Users/Hashim/Desktop/gettext/gettext-0.19.8'
make: *** [Makefile:369: all] Error 2
```

**Comment by tonytheodore** [comment]
Looks like it may be fiddly to build based on some of the patches and options in https://github.com/cygwinports/gettext.

The native gettext isn't required for all those packages, you could try:
```
make MXE_TARGETS="x86_64-w64-mingw32.static" -j$(nproc) MXE_USE_CCACHE= \
    libvpx ogg opus sdl2 theora vorbis x264
```
to make sure the toolchain and cross build is basically working. If it does, you can try "tricking" mxe into skipping the gettext build with an empty build rule (assuming you have the cygwin package installed):

```
make MXE_TARGETS="x86_64-w64-mingw32.static" -j$(nproc) MXE_USE_CCACHE= \
    gettext_BUILD_x86_64-unknown-cygwin= \ 
    lame libass libvpx ogg opus sdl2 theora vorbis x264 
```

**Comment by Kaos-Industries** [comment]
> 
> 
> Looks like it may be fiddly to build based on some of the patches and options in https://github.com/cygwinports/gettext.
> 
> The native gettext isn't required for all those packages, you could try:
> 
> ```
> make MXE_TARGETS="x86_64-w64-mingw32.static" -j$(nproc) MXE_USE_CCACHE= \
>     libvpx ogg opus sdl2 theora vorbis x264
> ```
> 
> to make sure the toolchain and cross build is basically working. If it does, you can try "tricking" mxe into skipping the gettext build with an empty build rule (assuming you have the cygwin package installed):
> 
> ```
> make MXE_TARGETS="x86_64-w64-mingw32.static" -j$(nproc) MXE_USE_CCACHE= \
>     gettext_BUILD_x86_64-unknown-cygwin= \ 
>     lame libass libvpx ogg opus sdl2 theora vorbis x264 
> ```

I'll try this now, but this does seem like a bit of a hack. Is it out of the question for the project to apply the `cygports` patches?

**Comment by Kaos-Industries** [comment]
So after a day or so of testing, here's what I've concluded. 

Running the command:

```
cd / && rm -rf /mxe && 
git clone https://github.com/mxe/mxe.git && cd /mxe &&
make MXE_TARGETS="x86_64-w64-mingw32.static" -j$(nproc) MXE_USE_CCACHE= \
libvpx ogg opus sdl2 theora vorbis x264
```

...builds without issues (apart from something that says there's nothing to be done for vorbis, which I found strange).

However, trying to then compile FFmpeg (which is my end goal here) with the following `configure` command:

```
cd /mxe/pkg && rm -rf ffmpeg &&
wget -O ffmpeg-snapshot.tar.bz2 https://ffmpeg.org/releases/ffmpeg-snapshot.tar.bz2 &&
tar xvf ffmpeg-snapshot.tar.bz2 1> /dev/null && rm -f ffmpeg-snapshot.tar.bz2 && cd ffmpeg &&
export PKG_CONFIG_PATH_x86_w64_mingw32_static=/mxe/usr/x86_64-w64-mingw32.static/lib/pkgconfig &&
export PKG_CONFIG_PATH=/mxe/usr/x86_64-w64-mingw32.static/lib/pkgconfig &&
./configure \
--extra-cflags="-static" \
--extra-ldflags="-static" \
--extra-libs="-pthread -lm" \
--prefix="/mxe/usr/x86_64-w64-mingw32.static" \
--pkg-config="pkg-config" \
--pkg-config-flags="--static" \
--enable-libopus \
--enable-libvorbis \
--enable-libvpx \
--enable-libx264 \
--enable-gpl \
--enable-nonfree \
--enable-static 
```

...always fails with some variation of:

ERROR: opus not found using pkg-config

...where *package* is whatever the first package being installed is.

Here's the config.log from ffmpeg/ffbuild :

https://gist.github.com/Kaos-Industries/ec62aa8358746e7c49267151f53c71d4

Running your second command:

```
cd / && rm -rf /mxe && 
git clone https://github.com/mxe/mxe.git && cd /mxe &&
make MXE_TARGETS="x86_64-w64-mingw32.static" -j$(nproc) MXE_USE_CCACHE= \
gettext_BUILD_x86_64-unknown-cygwin= \
lame libass libvpx ogg opus sdl2 theora vorbis x264 
```

...fails to compile glib, although based on the log it's possible this could be solved by making warnings not count as errors, which I'm not sure how to do with MXE.

Here's the log for that command:

https://gist.github.com/Kaos-Industries/953f5413c92c13c05426539b0012bc20

In the meantime, I've also had a chance to have a fresh look at compiling FFmpeg on Cygwin using its own MinGW-w64 toolchain (via the setup/package manager) and I realised that what I first thought was a PKG_CONFIG_PATH error was actually just an error with three packages in particular - libvorbis, libass, and libfreetype2 - and I managed to cross-compile a version of FFmpeg without them using that toolchain. 

Maybe those packages weren't working because Cygwin's mingw64 versions of those three packages are old or otherwise incompatible, I don't know, but I'm gonna have a go at compiling them packages from source, and if I'm successful, it might make much more sense for me to go down that route instead of continuing to use MXE, especially if errors like the above persist. If I do decideto go down that route, however, I'd still be happy to help you debug this issue, as I really appreciate MXE as a project and think it's necessary in a world where most of its direct competitors are outdated, clunky, or otherwise just suffer from obstinate development. Even if not in the immediate future for this particular project, I can definitely see it becoming useful in the future.

**Comment by tonytheodore** [comment]
> nothing to be done for vorbis, which I found strange

vorbis is built as a dep of theora so make finds there's "nothing to do" for that goal.

Your ffmpeg configure options don't actually use the mxe compiler, look at https://github.com/mxe/mxe/blob/master/src/ffmpeg.mk#L30-L33 for some of the options. This is a custom script, for autotools, setting `--host` is usually enough.

I've [added `ffmpeg_CONFIGURE_OPTS`](https://github.com/mxe/mxe/commit/4d8d9b32c92e3eb4a3656c3b33c4d93ec7208287) so you can try:
```
make MXE_TARGETS="x86_64-w64-mingw32.static" \
    -j$(nproc) \
    MXE_USE_CCACHE= \
    ffmpeg \
    ffmpeg_DEPS="\
        cc \
        bzip2 \
        libvpx \
        ogg \
        opus \
        sdl2 \
        theora \
        vorbis \
        x264 \
        yasm \
        zlib" \
    ffmpeg_CONFIGURE_OPTS="\
        --disable-gnutls \
        --disable-libass \
        --disable-libbluray \
        --disable-libbs2b \
        --disable-libcaca \
        --disable-libmp3lame \
        --disable-libopencore-amrnb \
        --disable-libopencore-amrwb \
        --disable-libspeex \
        --disable-libvidstab \
        --disable-libvo-amrwbenc \
        --disable-libvpx \
        --disable-libxvid"
```
I've tested it here (on darwin) and the reduced deps match the disables so you won't see any "not found" errors. You can also do:
```
make update-package-ffmpeg ffmpeg_UPDATE='echo snapshot'
```
or try a [plugin](https://github.com/mxe/mxe/tree/master/plugins) with [minimal deps](https://github.com/mxe/mxe/blob/master/plugins/examples/custom-qt-min/overrides.mk) to have a persistent local set of overrides.

> fails on compiling glib

This is another native build issue and is the [single hardest problem](https://github.com/mxe/mxe/issues/145) for mxe to solve. `glib` needs (very close, if not the same version) native tools to cross-build it's libraries and there's no way we can rely on package managers - sometimes they're too new, sometimes too old.

We're happy to accept patches from cygports, mxe automatically includes some [platform specific overrides](https://github.com/mxe/mxe/tree/master/plugins/native) so I'm pretty sure it could be made to work if you have the time and patience.

**Comment by tonytheodore** [comment]
If you use the snapshot, add `ffmpeg_SUBDIR=ffmpeg` to the `make` invocation since the subdirectory isn't versioned.

**Comment by tonytheodore** [comment]
I've added the [`cygwin` plugin directory](https://github.com/mxe/mxe/tree/master/plugins/native/cygwin) with some notes and gettext patches from cygwinports. Review the [cygports configure options](https://github.com/cygwinports/gettext/blob/master/gettext.cygport) and copy whatever you need into the [mxe gettext makefile](https://github.com/mxe/mxe/blob/master/plugins/native/cygwin/gettext.mk) (see the [darwin build rule](https://github.com/mxe/mxe/blob/master/plugins/native/darwin/gettext.mk) as an example.

You can use this as a starting point to get cygwin fully supported in mxe.

**Comment by Kaos-Industries** [comment]
> Your ffmpeg configure options don't actually use the mxe compiler, look at https://github.com/mxe/mxe/blob/master/src/ffmpeg.mk#L30-L33 for some of the options. This is a custom script, for autotools, setting `--host` is usually enough.
> 
> I've [added `ffmpeg_CONFIGURE_OPTS`](https://github.com/mxe/mxe/commit/4d8d9b32c92e3eb4a3656c3b33c4d93ec7208287) so you can try:
> 
> ```
> make MXE_TARGETS="x86_64-w64-mingw32.static" \
>     -j$(nproc) \
>     MXE_USE_CCACHE= \
>     ffmpeg \
>     ffmpeg_DEPS="\
>         cc \
>         bzip2 \
>         libvpx \
>         ogg \
>         opus \
>         sdl2 \
>         theora \
>         vorbis \
>         x264 \
>         yasm \
>         zlib" \
>     ffmpeg_CONFIGURE_OPTS="\
>         --disable-gnutls \
>         --disable-libass \
>         --disable-libbluray \
>         --disable-libbs2b \
>         --disable-libcaca \
>         --disable-libmp3lame \
>         --disable-libopencore-amrnb \
>         --disable-libopencore-amrwb \
>         --disable-libspeex \
>         --disable-libvidstab \
>         --disable-libvo-amrwbenc \
>         --disable-libvpx \
>         --disable-libxvid"
> ```
> 
> I've tested it here (on darwin) and the reduced deps match the disables so you won't see any "not found" errors. You can also do:
> 
> ```
> make update-package-ffmpeg ffmpeg_UPDATE='echo snapshot'
> ```
> 
> or try a [plugin](https://github.com/mxe/mxe/tree/master/plugins) with [minimal deps](https://github.com/mxe/mxe/blob/master/plugins/examples/custom-qt-min/overrides.mk) to have a persistent local set of overrides.
> 
> > fails on compiling glib
> 
> This is another native build issue and is the [single hardest problem](https://github.com/mxe/mxe/issues/145) for mxe to solve. `glib` needs (very close, if not the same version) native tools to cross-build it's libraries and there's no way we can rely on package managers - sometimes they're too new, sometimes too old.

So what exactly does this mean - that it will be impossible to get a version of MXE working with glib2.0 or Cygwin, or that it can eventually be made possible with work on the patches?

> I've added the cygwin plugin directory with some notes and gettext patches from cygwinports. Review > the cygports configure options and copy whatever you need into the mxe gettext makefile (see the  darwin build rule as an example).

> You can use this as a starting point to get cygwin fully supported in mxe.

Unfortunately I've had a look at gettext.cygport and it doesn't really look like I'll know what to do with that to get this working - what exactly I'd need to copy into the mxe gettext makefile. I have no actual programming experience whatsoever and I'm new to compiling and open source as a whole.

What I know I can do is run the mxe command you've provided though, if that will help. I assume this will 
build the dependencies and then build FFmpeg on its own without me needing to specify a configure separately?

To make sure you're aware, I did manage to get FFmpeg working with the packages that I wanted to using the MinGW-w64 toolchain and compiling most packages from source individually - so if I was the only reason for getting MXE to work on Cygwin you don't need to worry about that. If it's a lot of work for something that's relatively little priority for you, I understand that - on the other hand, if it's something of priority for you personally then I'm happy to do what I can do to help with that.

**Comment by Kaos-Industries** [comment]
Great news - the above command completely successfully, and I now have a working version based . I'm gonna try and build a version without

Something I've noticed - the project doesn't handle memory very well. Every time I've run the command, both without and without building FFmpeg, my RAM usage peaked to around 14GB. I only noticed it when I noticed performance issues, and this was while leaving `mxe` running overnight and long after it had finiished, so it's both responsible for the vast majority of this RAM and doesn't release it after it's finished building. That's something that definitely needs to be fixed, as if I didn't have 16GB of RAM installed on this machine it would have resulted in a BSOD.
